northstar/src/runtime/process/init.rs

```diff
@@ -40,11 +36,130 @@ pub(super) struct Init {
+    }
+    let (tx, rx) = mpsc::channel();
+
+    // This thread forwards the commands coming from the runtime
+    let command_tx = tx.clone();
+    std::thread::spawn(move || loop {
+        let command = rt_rx
+            .recv()
+            .expect("Failed to receive init command from runtime");
+        command_tx.send(Event::Command(command)).unwrap();
+    });
+
+    // This thread sends the exit status of the child processes
+    let exit_status_tx = tx;
+    let reaper_thread = std::thread::spawn(move || {
+        loop {
+            match waitpid(None, None) {
+                Ok(WaitStatus::Exited(pid, status)) => {
+                    exit_status_tx
+                        .send(Event::Exit {
+                            pid: pid.as_raw(),
+                            exit_status: ExitStatus::Exit(status),
+                        })
+                        .unwrap();
+                }
+                Ok(WaitStatus::Signaled(pid, status, _)) => {
+                    exit_status_tx
+                        .send(Event::Exit {
+                            pid: pid.as_raw() as i32,
+                            exit_status: ExitStatus::Signalled(status as u8),
+                        })
+                        .unwrap();
+                }
+                Ok(WaitStatus::Continued(_)) | Ok(WaitStatus::Stopped(_, _)) => continue,
+                Err(Errno::ECHILD) => {
+                    // wait till a child process is spawned
+                    // unparked by the consumer thread after a Exec is received
+                    std::thread::park();
+                }
+                e => panic!("Failed to waitpid: {:?}", e),
+            }
+        }
+    });
+
+    // Consume messages from threads
+    loop {
+        match rx.recv() {
+            Ok(Event::Exit { pid, exit_status }) => {
+                rt_tx
+                    .send(InitMessage::Exit { pid, exit_status })
+                    .expect("Failed to send exit status");
+
+                // Note:
+                // Currently the init process exists after the first child
+                // process exits.
+                exit(0);
+            }
+            Ok(Event::Command(InitApi::Exec { path, args, env })) => {
+                // Start new process inside the container
+                let child_pid = fork(|| {
+                    // Set seccomp filter
+                    if let Some(ref filter) = seccomp {
+                        filter.apply().expect("Failed to apply seccomp filter.");
+                    }
+
+                    // Checkpoint fds are FD_CLOEXEC and act as a signal for the launcher that this child
+                    // is started. Therefore no explicit drop (close) of _checkpoint_notify is needed here.
+                    panic!(
+                        "Execve: {:?} {:?}: {:?}",
+                        &path,
+                        &args,
+                        unistd::execve(&path, &args, &env)
+                    )
```

Comment:
is it correct that this panic will never execute if everything works as expected?

Reply:
Yes.

Reply:
yes, it was a bit confusing for me too :sweat_smile: but after the `execve` is executed this is a different program and that `panic` should not be trigger unless something goes wrong.